features/poll/impl/src/main/kotlin/io/element/android/features/poll/impl/create/CreatePollPresenter.kt

```diff
@@ -0,0 +1,168 @@
+
+import androidx.compose.runtime.Composable
+import androidx.compose.runtime.MutableState
+import androidx.compose.runtime.derivedStateOf
+import androidx.compose.runtime.getValue
+import androidx.compose.runtime.mutableStateOf
+import androidx.compose.runtime.remember
+import androidx.compose.runtime.rememberCoroutineScope
+import androidx.compose.runtime.saveable.Saver
+import androidx.compose.runtime.saveable.rememberSaveable
+import androidx.compose.runtime.setValue
+import dagger.assisted.Assisted
+import dagger.assisted.AssistedFactory
+import dagger.assisted.AssistedInject
+import io.element.android.libraries.architecture.Presenter
+import io.element.android.libraries.matrix.api.poll.PollKind
+import io.element.android.libraries.matrix.api.room.MatrixRoom
+import io.element.android.services.analytics.api.AnalyticsService
+import kotlinx.collections.immutable.ImmutableList
+import kotlinx.collections.immutable.toImmutableList
+import kotlinx.coroutines.launch
+import timber.log.Timber
+
+private const val MIN_ANSWERS = 2;
+private const val MAX_ANSWERS = 20;
+private const val MAX_ANSWER_LENGTH = 240;
+private const val MAX_SELECTIONS = 1;
+
+class CreatePollPresenter @AssistedInject constructor(
+    private val room: MatrixRoom,
+    private val analyticsService: AnalyticsService,
+    @Assisted private val navigateUp: () -> Unit,
+    // private val messageComposerContext: MessageComposerContext, // TODO
+) : Presenter<CreatePollState> {
+
+    @AssistedFactory
+    interface Factory {
+        fun create(backNavigator: () -> Unit): CreatePollPresenter
+    }
+
+    @Composable
+    override fun present(): CreatePollState {
+
+        var question: String by rememberSaveable { mutableStateOf("") }
+        var answers: List<String> by rememberSaveable() { mutableStateOf(listOf("", "")) }
+        var pollKind: PollKind by rememberSaveable(saver = pollKindSaver) { mutableStateOf(PollKind.Disclosed) }
+        var showConfirmation: Boolean by rememberSaveable { mutableStateOf(false) }
+
+        val canCreate: Boolean by remember { derivedStateOf { canCreate(question, answers) } }
+        val canAddAnswer: Boolean by remember { derivedStateOf { canAddAnswer(answers) } }
+        val immutableAnswers: ImmutableList<Answer> by remember { derivedStateOf { answers.toAnswers() } }
+
+        val scope = rememberCoroutineScope()
+
+        fun handleEvents(event: CreatePollEvents) {
+            when (event) {
+                is CreatePollEvents.Create -> scope.launch {
+                    if (canCreate) {
+                        room.createPoll(
+                            question = question,
+                            answers = answers,
+                            maxSelections = MAX_SELECTIONS,
+                            pollKind = pollKind,
+                        )
+                        // analyticsService.capture(PollCreate()) // TODO: Send PollCreate analytics.
+                        navigateUp()
```

Comment:
Is there a risk that calling this will cancel the poll creation (since the scope will be cancelled)?

Also we generally let the Node handle the navigation, so better to follow the existing pattern. The presenter set a boolean to the State that the flow is over, and the View communicate to the Node to handle the navigation.

Reply:
> Is there a risk that calling this will cancel the poll creation (since the scope will be cancelled)?

Yes there is, though generally the rust sdk is very quick and I never saw it happen. In general I think this might be a bigger architectural problem as all our presenters use `rememberCoroutineScope()`.

I've seen another case where a custom navigator has been used (`MessagesNavigator`) so I thought this approach would be better, the added benefit is that it is testable.